Add TFA.me integration to Home Assistant

DrMatthiasBlaschke opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds the new `tfa_me` integration to Home Assistant.

The integration provides sensor entities for TFA.me weather stations and allows
local data retrieval without relying on any cloud services.
It is based on the external Python library `tfa-me-ha-local`, which handles
device discovery and communication.

The integration follows the standard Home Assistant patterns and includes
full test coverage for configuration flows, coordinators and sensor entities.

External `tfa-me-ha-local` library at pypi:
https://pypi.org/project/tfa-me-ha-local/

Source code and bug trackers are available on GitHub at:
Source: "https://github.com/DrMatthiasBlaschke/tfa_me_ha_local"
Issues: "https://github.com/DrMatthiasBlaschke/tfa_me_ha_local/issues"

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes # none
- This PR is related to issue: none
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/40974
- Link to developer documentation pull request: N/A
- Link to frontend pull request: N/A

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][_integrations/tfa_me.markdown]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

DrMatthiasBlaschke commented on homeassistant/components/a_tfa_me_1/coordinator.py:
```diff
@@ -0,0 +1,101 @@
+        _LOGGER.info(msg)
+        try:
+            async with aiohttp.ClientSession() as session:
+                async with asyncio.timeout(5):  # 10 seconds timeout
+                    async with session.get(url) as response:
+                        if response.status != 200:
+                            raise UpdateFailed(f"HTTP Error {response.status}")  # noqa: TRY301
+
+                        # Get JSON reply from response
+                        json_data = await response.json()
+
+                        # Parse JSON data
+                        # sensors_new = {}
+                        for sensor in json_data.get("sensors", []):
+                            sensor_id = sensor["sensor_id"]
+
+                            for measurement, values in sensor.get(
+                                "measurements", {}
+                            ).items():
+                                entity_id = (
+                                    f"sensor.{sensor_id}_{measurement}"  # Entity ID
+                                )
+                                parsed_data[entity_id] = {
+                                    "sensor_id": sensor_id,
+                                    "sensor_name": sensor["name"],
+                                    "measurement": measurement,
+                                    "value": values["value"],
+                                    "unit": values["unit"],
+                                    "timestamp": sensor.get("timestamp", "unknown"),
+                                }
```
Oh, sorry, I didn't know that. This is my first integration.
A few quick questions before I waste more of your time reviewing:
- Is there a reference to this on the HA Developer pages? Link?
- What exactly isn't allowed to be used:
a) Only session = async_get_clientsession() .....?
b) Can the library on PyPI return JSON data, and can the integration then parse it?
c) Querying JSON data via URL is surely a common practice. Is there a library that can be easily used, or does each integration require its own library? It's just a really simple query: "IP/URL" sent and JSON data returned".

Thank you for your support and patience!

joostlek reviewed (changes requested):
1. Why do we change the copilot instructions?
2. Why is the domain `a_tfa_me_1`, that sounds weird
3. We should only have one platform in the initial PR
4. The HACS json is still there
5. Users shouldn't use the UI to set a custom polling interval
6. Please look at other config flows how they do it

DrMatthiasBlaschke commented on homeassistant/components/tfa_me/__init__.py:
```diff
@@ -0,0 +1,98 @@
+    delta_interval = timedelta(seconds=LOCAL_POLL_INTERVAL)
+
+    # Use multiple entities
+    multiple_entities = entry.data[CONF_MULTIPLE_ENTITIES]
```
- We thought so too at first, but 95% or more of users only have one station. This makes the station ID unnecessary.
- With station ID included we get a flood of entities and if the reception quality is good from all stations, it's unnecessary to create multiple entities. For some users, it's sufficient to have the measured value once, regardless of which station it comes from.
- And the worst thing is: The entity names are very long and are often cut off on the HA UI. This makes selection and overview very difficult.
Therefore, we prefer that individual users be able to configure what and how they want to use entities (single, multiple). We don't want to dictate anything to the user, but rather offer maximum flexibility. Should this lead to problems in practice or a lot of support inquiries, we can always consider changing something. At the moment, we consider the long entity names to be a much bigger problem.

joostlek replied:
> Should this lead to problems in practice or a lot of support inquiries, we can always consider changing something. At the moment, we consider the long entity names to be a much bigger problem.

A long entity name is probably another issue. But having duplicate IDs is a hard no-go.

So if I understand correctly, the device we connect to is a receiver for data, and 2 receivers can receive the same data from the same device?

DrMatthiasBlaschke commented on homeassistant/components/tfa_me/__init__.py:
```diff
@@ -0,0 +1,98 @@
+    entry.async_on_unload(entry.add_update_listener(async_update_listener))
+
+    # Save coordinator for later usage
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```
Bug, correct once is enough, will be removed.

DrMatthiasBlaschke commented on homeassistant/components/tfa_me/config_flow.py:
```diff
@@ -0,0 +1,247 @@
+        opt_dict = [
+            SelectOptionDict(value="none", label="None"),
+            SelectOptionDict(value="discover_sensors", label="Discover new sensors"),
+            SelectOptionDict(value="action_rain", label="Reset all rain sensors"),
```
The rain sensors only provide the raw data. This value can be a few hundred or a few thousand millimeters (mm), depending on how long the sensor has been running and how long it has rained. This value is not very meaningful, since neither the user nor the integration knows how long it has been running.
The integration therefore contains a mini-message queue in which the measured values ​​from the last hour/last 24 hours are stored, including a UTC timestamp. The integration calculates the correct values ​​from this. Don't worry: These are only a few values ​​and should only take up a small amount of RAM. The user doesn't have to do anything themselves.
And yes: If the integration or HA is restarted, the values ​​disappear (back to 0.0 mm). This cannot be changed. But that doesn't matter. It's also impossible to say when it rained if HA is offline. We then have no history.
<img width="369" height="378" alt="rain_sensor_entities" src="https://github.com/user-attachments/assets/94b37ba5-82fd-4a9e-bb47-a8141c0ca71e" />

joostlek replied:
> The rain sensors only provide the raw data. This value can be a few hundred or a few thousand millimeters (mm), depending on how long the sensor has been running and how long it has rained. This value is not very meaningful, since neither the user nor the integration knows how long it has been running.

Why would that matter? On one hand you mention it can't store data, on the other hand its a realtime data?

DrMatthiasBlaschke replied:
I'm not sure if I understand this correctly:
The meter reading in the rain sensor is an absolute value that is continuously incremented. This value is correct, but not particularly useful for the customer. For example, if I look at the meter reading in the morning and again in the evening, I have to remember/write down the reading from that morning to see how much it rained during that period. This is extremely impractical. Therefore, the integration calculates two additional values: rain in the last hour and the last 24 hours. This is much more meaningful. You can also use these values ​​directly for automation. For example, if it hasn't rained enough in the last 24 hours, you can activate a pump to water something.

DrMatthiasBlaschke commented:
Code changes:
- Removed comments with incorrect style
- Moved some functions to PyPI Lib
- Removed unnecessary options in config_flow/options_flow, as they can also be called via existing HA UI functions (reloading integration, updating entities)
- Removed unused functions
- Adapted test functions to code changes

joostlek reviewed (changes requested):
In some comments you mention you agree with the suggestion but you didn't change it. Please resolve any conversation you think you solved correctly, but keep open ongoing discussions

DrMatthiasBlaschke commented:
> In some comments you mention you agree with the suggestion but you didn't change it. Please resolve any conversation you think you solved correctly, but keep open ongoing discussions

-> No did not agree to remove the option "multiple ID." Always using the station ID creates entity names that are too long and unreadable, as the UI often truncates them. For 95% or more of all users, this doesn't matter at all, as they only have one station. Incidentally, HA doesn't create entities twice! If an entity name already exists and it's "added" again, HA doesn't take over. It's discarded. I've tested this with multiple stations, and nothing crashes or aborts.

But: If you want, I can of course also check in the data model whether an entity name already exists before adding it. It amounts to the same thing, but if it's a better approach, I'm fine with it.

DrMatthiasBlaschke commented:
> > Depending on the width of the browser screen, it can be difficult to display the entire name in lists. This problem cannot be solved, as the user can always configure the browser in such a way that a meaningful display is no longer possible. However, if you view the entity in its individual view, everything is always readable.
> 
> Again: you're looking at a bug. Your entity names shouldn't be that long. This means you're doing something wrong. Your unique id doesn't influence your name unless you did something wrong.

>> Could it be that we're completely talking past each other? 
What is the maximum length of an entity name added with async_add_entities()?
I find names e.g. for weather forecast  like: "sun_next_midnight_2" (19 character) and my is "a012345678_humidity" (18 characters). So why my names are to long and forecasts are not ?

> 
> > But it's not a bug that HA doesn't allow duplicate entity IDs, is it? If an ID already exists and you call async_add_entities(), HA generates a message on the console: "This entitiy already exists..." That's absolutely correct. The entity isn't added again. It works like a duplicate filter.
> 
> It is. Yes it works today, but we need to think about the future. We're not logging it for shits and giggles.

>> OK, got it. I will add a function that verifies checks whether entities names already exist BEFORE async_add_entities() is called. This eliminates the error message. This certainly makes more sense. OK?

DrMatthiasBlaschke commented:
> Yes that is fine as a name, you should use `_attr_has_entity_name = True` and it will prepend the device name.
> 
> But IMO we should add the station ID to the unique id. But your counter argument always was that the unique id is too long and thus the name would be too long as well and that it would be meh

I've been experimenting a bit more with entities and the registry. At runtime, it's no problem to detect duplicate entity IDs that have already been registered by other instances of the integration. That works well. However, there are problems when restarting HA. The entities are already present at that point; they aren't created anew, but only updated. This can lead to duplicate entities names appearing on the first retrieval of the current data, which is obviously unacceptable.

Therefore, the only pragmatic solution is to use the gateway and sensor ID as the entity names. This guarantees that no duplicate IDs can arise in the system.

Long sensor names are also not a problem after consultation with our client TFA. The user can optionally display the gateway ID in the sensor name. However, this is only for differentiation during the initial installation. Typically he will not use "TFA.me A01234567 (012345678)" or "TFA.me A01234567" but will rename it to "Living Room", "Kitchen", "Garage", etc.

joostlek reviewed (changes requested):
There's a merge conflict and some things are still not updated as you said it was, so please re-review

mik-laj commented on homeassistant/components/tfa_me/__init__.py:
```diff
@@ -0,0 +1,70 @@
+async def async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
+    """Will be called when options are changed via UI."""
+
+    reset_rain = entry.options.get("action_rain", False)
```
Should we use new value in some way?

DrMatthiasBlaschke commented on homeassistant/components/tfa_me/__init__.py:
```diff
@@ -0,0 +1,70 @@
+    return unload_ok
+
+
+async def async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
```
ok, if this also will work no problem.

DrMatthiasBlaschke replied:
Removing it is ok, this has no negative impact on functionality.

joostlek reviewed (changes requested):
Marking it as draft as some things still haven't been adressed and there are still ongoing discussions

DrMatthiasBlaschke commented:
> Marking it as draft as some things still haven't been adressed and there are still ongoing discussions

OK.

I've reviewed the request changes again and hope I haven't overlooked anything. They are:
- Redundant comments removed
- Unused/unnecessary functions removed
- Entities are unique; duplicates are no longer possible
- Error messages are now coming from the translation key

What else is missing or needs further discussion?
Please let me know. Thank you!

joostlek reviewed (changes requested):
I am going to push back again. Foremost that if you open the changed files, you will immediately see something that we spoke about, you agreed about and yet you didn't change it, or at least, you did change something now, but this is still not what we expect.

Other than that, there are some things in the init.py which you acknowledged but didn't change and the rain sensor reset should not be needed.

DrMatthiasBlaschke commented:
> > Resetting the rain sensor is absolutely necessary. I have explained this in detail
> 
> Yes and I explained that we have device classes for this that take this into account. See [#152802 (comment)](https://github.com/home-assistant/core/pull/152802#discussion_r2448080773). Except there was no response on this. In [#152802 (comment)](https://github.com/home-assistant/core/pull/152802#discussion_r2432576046) I clearly stated the line that we don't allow such abstractions to exist in Home Assistant. Perhaps I could have been clearer in this occasion, but so far this comment has been ignored and now when bringing it up you still push back.
> 
> > You don't have the right to negatively impact our product features just because you lack the necessary expertise. This has nothing to do with a review; it's an unacceptable interference with the product's functionality.
> 
> Yes, I don't know everything. My task is to ask the right questions and wonder why we do the things we do and to make sure the choices make sense. Making sense also means that they follow the Home Assistant rules. There are several weather stations integrations in Home Assistant and yet none of them have a reset rain flow, so why should we have it here? Again, we don't want abstractions like these in Home Assistant. There is some room in how you implement certain things in Home Assistant, but again, if they make sense, not because "we think people want to reset these values using a flow that you can't automate". No, that's not the Home Assistant way of solving these things, which I am trying to guide you towards, but it feels like I am running into a brick wall.
> 
> > The init.py file is 55 lines long. And there's absolutely nothing in it that was addressed. If I remove that code there, it no longer works correctly.
> 
> I mean, I've stated that you are storing the coordinator twice and you acknowledged it and did not change it. See [#152802 (comment)](https://github.com/home-assistant/core/pull/152802#discussion_r2431589790).
> 
> It's very frustrating for me to have guided multiple times on what files not to edit, only to open the changed files to see, these files having changes. <img alt="image" width="1511" height="533" src="https://private-user-images.githubusercontent.com/7083755/512142841-a0180615-7ca8-492b-b5e9-9a2bbbdb414f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI4NTMwOTEsIm5iZiI6MTc2Mjg1Mjc5MSwicGF0aCI6Ii83MDgzNzU1LzUxMjE0Mjg0MS1hMDE4MDYxNS03Y2E4LTQ5MmItYjVlOS05YTJiYmJkYjQxNGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTExMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMTFUMDkxOTUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTk5ZDhjOTgxODVlNWY4MTZjYWNiYmYzOGU4YzcwNTJlNjlhMjI2ZThiZTlkODQxMTg0ZGEzOGU5NjA3NjhhNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.wrm4A735JDtlA5FjClw1QrIy_GIl-NVE2gEb-C_II_U">
> 
> Keep in mind I am asking questions with a reason. To ask the question more concrete: what part of my comments is not clear to you?

Yes, you're right, I didn't reply to that, sorry.
Here's the explanation:

Back to the setup of the rain sensor. It contains a small tipping bucket with two water reservoirs. Every time a certain amount of rain falls into the reservoir, the tipping bucket tips over and the counter reading is incremented. Each tipping of the bucket corresponds to 0.x mm of rain, meaning the counter reading jumps by 0.x mm.

The correct graphical representation is therefore a kind of stepped function, like a staircase.
It is also important that the graphical representation corresponds to the time course of the measured value.
This shows how much rain has fallen in the period x (1 hour, 24 hours, since HA start, etc.).
And the user wants to be able to restart these measurements at any time; therefore, teh reset option is needed.
Attached is a snapshot. This is how it should look:

<img width="550" height="344" alt="rain_relative_as_stair_function" src="https://github.com/user-attachments/assets/70cbb745-21a8-4a73-a5db-72a44b11e677" />

If we use SensorStateClass.TOTAL or SensorStateClass.TOTAL_INCREASING, it looks completely different. The graph shows when and how much rain fell (combined/aggregated over a specific period). 
This contains two errors:
- The graphical representation does not correspond to the desired value (rain x hours).
- And if values ​​are being aggregated, then they should be bar charts, according to the measurement principle, like here:
 
![rain_bar_chart](https://github.com/user-attachments/assets/425464e1-ee74-4901-ba75-4eba808c0f7d)

- But HA interpolates the values in another way. To me, this looks like "cubic splines". The curves do pass through the measured values, but the curve representation is mathematically incorrect (even if it looks nice) and this graphical representation is also not desired. Therefore, it cannot be used here.

<img width="545" height="340" alt="rain_splines_2" src="https://github.com/user-attachments/assets/844b402b-4436-4fd6-b330-d68755e2012b" />

I hope that explains it better. If you have any further questions, please let me know.

To init.py:
I think I've got it now. That was a misunderstanding on my part.
I saved it 3 times before:
- Empty/default: hass.data.setdefault(DOMAIN, {})
- Compatible with HA 23.6 and older: hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
- Modern: entry.runtime_data = coordinator

I only removed the first line.  But OK, if the old
- hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
code is no longer needed or should not be used, then of course I can remove it.

To copillot-instruction.md:

I still don't know why the original file isn't displayed there anymore, or which tool or script changed what. I'll try to find the original file. For now, you can leave PR on "Draft."

BTW: It's completely irrelevant which files were changed apart from my two integration folders. When the new integration is added, only two folders, ..components/tfa_me and test/.../tfa_me, are inserted. It's a simple copy and paste. I don't have the right to change other integrations or anything in the core. Therefore, it can simply be ignored. Only what's in the two new integration folders is relevant. I have exclusive access to these two folders and nobody else. Nothing from other developers is merged here. Conversely, it looks exactly the same. Therefore, I don't understand why this is even part of the review.

pantherale0 commented on homeassistant/components/tfa_me/__init__.py:
```diff
@@ -0,0 +1,55 @@
+    await coordinator.async_config_entry_first_refresh()
+
+    # Set coordinator data
+    entry.runtime_data = coordinator
```
I think we should change all of this to:


```suggestion
    entry.runtime_data = coordinator = TFAmeDataCoordinator(
        hass, entry, host, delta_interval, name_with_station_id
    )
    await coordinator.async_config_entry_first_refresh()
```

DrMatthiasBlaschke replied:
OK, will be changed/replaced

pantherale0 commented on homeassistant/components/tfa_me/__init__.py:
```diff
@@ -0,0 +1,55 @@
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id, None)
+    return unload_ok
```
```suggestion
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```

pantherale0 commented on homeassistant/components/tfa_me/config_flow.py:
```diff
@@ -0,0 +1,177 @@
+        self.data: dict[str, Any] = {}
+        self.name_with_station_id: bool = False
+
+    _LOGGER.debug("TFA.me config flow")
```
Wouldn't this create a runtime error since it is not in a callable function?

I would remove.

pantherale0 commented on homeassistant/components/tfa_me/config_flow.py:
```diff
@@ -0,0 +1,177 @@
+        # Error, validation failed
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors
+        )
```
I think this should be rewritten, you can tidy this up and remove duplicate function calls by opposing the logic on `user_input` (rather than checking if user_input is None, check that it is not None and perform all your validation / setup underneath.

Then the last piece of logic in the chain will always be to show the form.

pantherale0 commented on homeassistant/components/tfa_me/const.py:
```diff
@@ -0,0 +1,118 @@
+    "A5": "Sensor A5: T",
+    "A6": "Sensor Prof. A6: T/H",
+    # Add other sensors here ...
+}
```
```
# Add other sensors here ...
```

This could be dynamic, but I've not seen the rest of the code yet to understand what this is doing.

pantherale0 commented on homeassistant/components/tfa_me/coordinator.py:
```diff
@@ -0,0 +1,135 @@
+
+            # Convert JSON to TFA.me data for HA
+            parsed_data = tfa_me_data.json_to_entities(json_data=json_data)
+            self.gateway_id = tfa_me_data.get_gateway_id()
```
Your python module should not be aware of Home Assistant. Instead it should provide generic data classes & attributes that Home Assistant can access.

You can then use these attributes directly in lambda expressions in various platforms.

pantherale0 commented on homeassistant/components/tfa_me/coordinator.py:
```diff
@@ -0,0 +1,135 @@
+            _LOGGER.error(msg)
+            if self.first_init == 0:
+                raise ConfigEntryNotReady(f"Unexpected error: {err}") from err
+            raise UpdateFailed(f"Unexpected error: {err}") from err
```
Some of these do the same thing, so why not:

(or similar)
```suggestion
        except (TFAmeHTTPError, TFAmeJSONError) as err:
            # Device responding but data invalid
            _LOGGER.exception("Invalid response")
            raise UpdateFailed(f"Invalid response: {err}") from err

        except (TFAmeTimeoutError, TFAmeConnectionError, TFAmeException, Exception) as err:
            _LOGGER.exception("Error occurred")
            if self.first_init == 0:
                raise ConfigEntryNotReady(
                    f"Cannot reach {resolved_host}: {err}"
                ) from err
            raise UpdateFailed(f"Connection problem: {err}") from err
```

pantherale0 commented on homeassistant/components/tfa_me/manifest.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "domain": "tfa_me",
+  "name": "TFA.me station",
+  "after_dependencies": ["sensor", "hassio"],
```
This is not required

DrMatthiasBlaschke replied:
OK, will be removed

pantherale0 commented on homeassistant/components/tfa_me/sensor.py:
```diff
@@ -0,0 +1,481 @@
+    """TFA.me sensor entity, represents in HA a single measurement value of a sensor."""
+
+    _attr_has_entity_name = True
+    _attr_should_poll = True
```
Should it poll if the coordinator handles the updates?

pantherale0 commented on tests/components/tfa_me/test_additional.py:
```diff
@@ -0,0 +1,19 @@
+    assert validator.is_valid_ip_or_tfa_me("012-345-678")
+
+    # Invalid Host/IP
+    assert not validator.is_valid_ip_or_tfa_me(42)
```
Why are we testing library functions in HA? These belong in tests for that specific library.

pantherale0 commented on tests/components/tfa_me/test_coordinator.py:
```diff
@@ -0,0 +1,356 @@
+    assert "sensor.017654321_a12345678_rain" in result
+    assert "sensor.017654321_a12345678_rssi" in result
+    assert "sensor.017654321_a12345678_lowbatt" in result
+    assert "sensor.017654321_a12345678_lowbatt_txt" in result
```
We can use snapshot testing instead for the entities, but for just the coordinator, this should not be necessary.

pantherale0 commented on tests/components/tfa_me/test_init.py:
```diff
@@ -0,0 +1,76 @@
+        },
+        entry_id="1234",
+        unique_id="unique_1234",
+    )
```
Duplicate, as above, centralize all of these into conftest.

DrMatthiasBlaschke replied:
OK, all centralized in conftest

zweckj commented:
changes have been requested, setting to draft

DrMatthiasBlaschke commented on tests/components/tfa_me/test_config_flow.py:
```diff
@@ -0,0 +1,165 @@
+        )
+
+    assert result["type"] == data_entry_flow.FlowResultType.FORM
+    assert result["errors"]["base"] in ("cannot_connect", "unknown")
```
OK, will be rewritten.

DrMatthiasBlaschke commented on homeassistant/components/tfa_me/coordinator.py:
```diff
@@ -0,0 +1,135 @@
+        # Try to update data from station URL: e.g. "http://192.168.1.38/sensors"
+        try:
+            # New TFA.me client
+            tfa_me_client = TFAmeClient(resolved_host, "sensors", log_level=1)
```
OK. will be replaced by reusable session/client.

DrMatthiasBlaschke commented on homeassistant/components/tfa_me/data.py:
```diff
@@ -0,0 +1,27 @@
+        """Initialize client with host."""
+
+        if host == "":
+            raise TFAmeException("host_empty")
```
OK, I thought one had to/should generally address such a case. But it can be removed.

DrMatthiasBlaschke commented on homeassistant/components/tfa_me/sensor.py:
```diff
@@ -0,0 +1,481 @@
+    except Exception as error:
+        raise ConfigEntryNotReady(
+            f"Station not available: {error}"
+        ) from error  # Catch errors here
```
OK, will be removed

DrMatthiasBlaschke commented on homeassistant/components/tfa_me/sensor.py:
```diff
@@ -0,0 +1,481 @@
+    # Save function in Home Assistant so that it can be called as service
+    hass.data[DOMAIN][
+        entry.entry_id
+    ].async_discover_new_entities = async_discover_new_entities
```
OK, will be removed

pantherale0 commented on homeassistant/components/tfa_me/config_flow.py:
```diff
@@ -0,0 +1,177 @@
+        if user_input is not None:
+            if "select_option" in user_input:
+                if user_input["select_option"] == "action_rain":
+                    return await self.async_step_action_rain(user_input)
```
Regardless of the comment below, I don't think we need a options menu for just a single option?

pantherale0 commented on homeassistant/components/tfa_me/config_flow.py:
```diff
@@ -0,0 +1,177 @@
+        """Entry point for option: Reset all rain sensors."""
+        if user_input is not None:
+            if user_input["select_option"] == "action_rain":
+                coordinator = self.hass.data[DOMAIN][self.config_entry.entry_id]
```
Remove use of hass.data and instead rely on entry.runtime_data for consistency

pantherale0 commented on homeassistant/components/tfa_me/coordinator.py:
```diff
@@ -0,0 +1,179 @@
+                sensor_id = sensor["sensor_id"]
+
+                for m_name, values in sensor.get("measurements", {}).items():
+                    entity_id = f"sensor.{gateway_id}_{sensor_id}_{m_name}"
```
We shouldn't rely on this being the entity id as a user could change this. Same applies for the sensor name below.

Instead rely on the unique ID of the entity.

A better solution would be to just store the values without the knowing the context of the entity. For example:

```python
data = {
"wind_direction": value,
"rain_1_hour": value2,
}
```

Then use SensorEntityDescriptions to access the data.. For example:

https://github.com/home-assistant/core/blob/4960871c8488582d52187f25df61e73fe1994d36/homeassistant/components/nintendo_parental_controls/sensor.py#L26-L57

pantherale0 commented on homeassistant/components/tfa_me/data.py:
```diff
@@ -0,0 +1,68 @@
+
+    def _get_client(self) -> TFAmeClient:
+        """Return a TFA.me client instance for this host."""
+        if self._client is None:
```
Why would this be None?

pantherale0 commented on homeassistant/components/tfa_me/sensor.py:
```diff
@@ -0,0 +1,449 @@
+    """Set up TFA.me as Sensor."""
+
+    # Get coordinator
+    coordinator = hass.data.setdefault(DOMAIN, {})[entry.entry_id]
```
`entry.runtime_data`

DrMatthiasBlaschke committed:
Added zeroconf & zeroconf tests

zweckj reviewed (changes requested):
A pypi link is no link to your library (and the links to the library from pypi only yield 404). We require libraries to be open source and assinged with a OSI approved license as part of our `dependency-transparency` quality scale item. Please make sure the source code is publicly accessible and pushed to Pypi with tagged, automated releases. Thanks.

DrMatthiasBlaschke committed:
New coordinator data model, code simplified

DrMatthiasBlaschke committed:
Changed entity creation, renamed classes, updated tests

DrMatthiasBlaschke committed:
Removed OptionsFlow & ZeroConf for later PR

DrMatthiasBlaschke committed:
Various changes based on: @Copilot commented on this PR

DrMatthiasBlaschke committed:
Removed typos in tests

DrMatthiasBlaschke committed:
Bugfix: Missing enteties update after HA or/and
gateway restart, New:automatically add new sensors

DrMatthiasBlaschke committed:
Added comments regarding the entity IDs generated by HA.

DrMatthiasBlaschke committed:
Several changes concerning Copilot comments on PR

DrMatthiasBlaschke committed:
More changes based on: @Copilot commented on this PR

DrMatthiasBlaschke committed:
Restore missing files

DrMatthiasBlaschke committed:
Changed tests based on @Copilot commented on PR

DrMatthiasBlaschke committed:
Updated requirements_test_all

DrMatthiasBlaschke committed:
Much more changes based on @Copilot comments on PR

DrMatthiasBlaschke committed:
Updated requirements_test_all.txt

DrMatthiasBlaschke committed:
More changes based on: @Copilot comments on PR

DrMatthiasBlaschke committed:
Changes made based on errors in workflow run

DrMatthiasBlaschke committed:
Again more changes based on: @Copilot comments PR

DrMatthiasBlaschke committed:
Updated requirements_all

DrMatthiasBlaschke committed:
Some fixes based on: @Copilot comments on this PR

DrMatthiasBlaschke committed:
Updated quality_scale new docs-xxx entries bronze

DrMatthiasBlaschke committed:
Use tfa_me_ha_local 1.0.8, removed comments

DrMatthiasBlaschke committed:
Added/changed some tests @Copilot comments

DrMatthiasBlaschke committed:
Removed superfluous yaml, tests and units

DrMatthiasBlaschke committed:
Updated snapshots

DrMatthiasBlaschke committed:
Updated comments in quality scale & doc strings

DrMatthiasBlaschke committed:
Rebased & updated mdi_icons.py

DrMatthiasBlaschke committed:
Removed "sensor." from IDs, updated quality_scale

DrMatthiasBlaschke committed:
RSSI & low battery are EntityCategory.DIAGNOSTIC

DrMatthiasBlaschke committed:
Udpated snapshot

DrMatthiasBlaschke committed:
Updated tests, wind direction, typos
